### PR TITLE
16 connect ttn sensors to lorawan agent

### LINKED ISSUE #16: connect ttn sensors to lorawan agent

### PR DESCRIPTION
**Pull request by casarf**
## Proposed changes

Describe the big picture of your changes here to communicate to the
maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to
that issue.

## Types of changes

What types of changes does your code introduce to the project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Maintenance (update of libraries or other dependences)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING][contrib] doc
- [x] I have signed the [CLA][cla]
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have run **all** the existing tests locally (not just those related to my feature) and there are no errors
- [ ] After the last push to the PR branch, I have run the lint script locally and there are no changes to the code base
- [ ] I have updated the [RELEASE NOTES][release]
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

[cla]: https://raw.githubusercontent.com/orchestracities/lego-diorama-usi-2022/master/individual_cla.pdf
    "Martel Open Source Software Individual Contributor License Agreement"
[contrib]: https://github.com/orchestracities/lego-diorama-usi-2022/blob/master/CONTRIBUTING.md
    "Contributing to Lego Diorama"
[release]: https://github.com/orchestracities/lego-diorama-usi-2022/blob/master/RELEASE_NOTES.md
    "Lego Diorama Release Notes"

### CODE REVIEW

File: README_TTN.md [hunk 1]
```diff
@@ -103,6 +107,134 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+
+```bash
+curl --location --request POST 'localhost:4042/iot/services' \
+--header 'fiware-service: openiot' \
+--header 'fiware-servicePath: /' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+  "services": [
+    {
+      "entity_type": "waste",
```
Comment (chicco785) [inline README_TTN.md]: WasteContainer

File: README_TTN.md [hunk 2]
```diff
@@ -103,6 +107,134 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+      "resource": "8CF957200005727C",
+      "attributes": [
+        {
+          "object_id": "temperature_1",
+          "name": "temperature",
+          "type": "Number"
+        },
+        {
+          "object_id": "relative_humidity_1",
+          "name": "relative_humidity",
```
Comment (chicco785) [inline README_TTN.md]: relativeHumidity

File: README_TTN.md [hunk 3]
```diff
@@ -103,6 +107,134 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+          "type": "Number"
+          },
+          {
+            "object_id": "digital_in_1",
+            "name": "flame_detected",
```
Comment (chicco785) [inline README_TTN.md]: the thing is a bit more complex than this, you need to compute an attribute `status` based on the different sensor values.

ok. Container is where it must be and stands properly.
lidOpen. Container's lid has been opened and not closed after a certain amount of time.
dropped. Container has been dropped for some reason.
moved. Container has been moved from its regular position and has not come back.
vandalized. Container has been damaged or destroyed due to vandalism.
burning. Container is burning and an immediate action has to be taken.
unknown. The status of the container is not known to the system.

Comment (chicco785) [reply README_TTN.md]: so you need an named "status" an use expressions to compute the value. anyhow, check the old code:
https://gitlab.com/orchestracities/iot/-/blob/master/diorama/src/waste/create_device.sh

File: README_TTN.md [hunk 4]
```diff
@@ -103,6 +107,134 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+            "type": "Number"
+            },
+            {
+              "object_id": "digital_in_2",
+              "name": "distance",
```
Comment (chicco785) [inline README_TTN.md]: https://gitlab.com/orchestracities/iot/-/blob/master/diorama/src/waste/waste.ino#L118

this is a fillingLevel, so you need to compute it as %. you can compute it in arduino or here, if you compute it here, you need a jexl expression in definition: https://iotagent-node-lib.readthedocs.io/en/2.23.0/expressionLanguage.html

File: README_TTN.md (hunk 3 above)
Comment (chicco785) [reply README_TTN.md]: https://gitlab.com/orchestracities/iot/-/blob/master/diorama/src/waste/waste.ino#L143

if you want to compute the status in arduino, if not you can use a jexl expression here (may be more complex)

File: platform-stack/services [hunk 5]
```diff
@@ -115,8 +235,10 @@ case "${command}" in
 		${dockerCmd} up -d --remove-orphans
 		waitForMongo
 		addDatabaseIndex
+		# waitForIoTAgent
 		waitForOrion
 		displayServices
+		agentSubscription
```
Comment (chicco785) [inline platform-stack/services]: name it at least loraAgent

File: README_TTN.md [hunk 6]
```diff
@@ -103,6 +107,134 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+                },
+                "app_eui": "8CF957200005727C",
+                "application_id": "oc-diorama-001@ttn",
+                "application_key": "51F58CAC3F5735E3D1F88DD3EADBE9C6",
+                "data_model": "application_server"
+            }
+          }
+    },
+    {
+      "entity_type": "environment",
```
Comment (chicco785) [inline README_TTN.md:181]: see the old code:
https://gitlab.com/orchestracities/iot/-/blob/master/diorama/src/environment/create_device.sh

**Review by chicco785 — CHANGES_REQUESTED** [review]

File: README_TTN.md [hunk 7]
```diff
@@ -103,6 +107,139 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
```
Comment (chicco785) [inline README_TTN.md:181]: `AirQualityObserved`

File: README_TTN.md [hunk 8]
```diff
@@ -103,6 +107,139 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+          "name": "UVvalue",
+          "type": "Number"
+          }
+      ],
+      "internal_attributes": {
+        "lorawan": {
+          "application_server": {
+            "host": "eu1.cloud.thethings.network",
+            "username": "oc-diorama-002@ttn",
+            "password": "NNSXS.WYCQ2VM2FNMXR5DNYYCYAHAQAACG4XA3FTBMRYA.SNSYCYYNLGVM4POZWSQ4VINZOUUPT5OAV3DU436W6AZXPRTLQCZAC0MM3NT01T4L1ANO",
```
Comment (chicco785) [inline README_TTN.md]: never put secrets in a git repo.

File: README_TTN.md [hunk 9]
```diff
@@ -103,6 +107,139 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+              "name": "status",
+              "type": "Number"
+            }
+          ],
+          "internal_attributes": {
+            "lorawan": {
+              "application_server": {
+                "host": "eu1.cloud.thethings.network",
+                "username": "oc-diorama-001@ttn",
+                "password": "NNSXS.KQRRO4R7BXMHELHQZOMELYUJ2GQUU4S5EI2I6QY.P7FJVWQRODOLCOHOAC2GQMEDO65LINW6YAOBKL6YMUU2K5C4R5MAC0MM3NT01T4L1ANO",
```
Comment (chicco785) [inline README_TTN.md]: never put secrets in a public git repo

File: README_TTN.md [hunk 10]
```diff
@@ -103,6 +107,139 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+            "type": "Number"
+            },
+            {
+              "object_id": "digital_in_2",
+              "name": "fillingLevelPercentage",
+              "type": "Number"
+            },
+            {
+              "object_id": "accelerometer_1",
+              "name": "accelerometerXYZ",
```
Comment (chicco785) [inline README_TTN.md]: are these needed? isn't this now part of "status"?

File: README_TTN.md [hunk 11]
```diff
@@ -103,6 +107,139 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+              "type": "StructuredValue"
+            },
+            {
+              "object_id": "accelerometer_2",
+              "name": "gAccellerationXYZ",
```
Comment (chicco785) [inline README_TTN.md]: see above

File: platform-stack/services [hunk 12]
```diff
@@ -98,6 +98,135 @@ waitForIoTAgent () {
...
+	                    "name": "status",
+	                    "type": "Number"
+	                }
+	            ],
+	            "internal_attributes": {
+	                "lorawan": {
+	                    "application_server": {
+	                        "host": "eu1.cloud.thethings.network",
+	                        "username": "oc-diorama-001@ttn",
+	                        "password": "NNSXS.KQRRO4R7BXMHELHQZOMELYUJ2GQUU4S5EI2I6QY.P7FJVWQRODOLCOHOAC2GQMEDO65LINW6YAOBKL6YMUU2K5C4R5MA",
```
Comment (chicco785) [inline platform-stack/services]: see comment above

File: platform-stack/services [hunk 13]
```diff
@@ -98,6 +98,135 @@ waitForIoTAgent () {
...
+	                    "type": "Number"
+	                },
+	                {
+	                    "object_id": "digital_in_2",
+	                    "name": "fillingLevelPercentage",
+	                    "type": "Number"
+	                },
+	                {
+	                    "object_id": "accelerometer_1",
+	                    "name": "accelerometerXYZ",
```
Comment (chicco785) [inline platform-stack/services]: see comment above

File: platform-stack/services [hunk 14]
```diff
@@ -98,6 +98,135 @@ waitForIoTAgent () {
...
+	                    "type": "StructuredValue"
+	                },
+	                {
+	                    "object_id": "accelerometer_2",
+	                    "name": "gAccellerationXYZ",
```
Comment (chicco785) [inline platform-stack/services]: see comment above

File: platform-stack/services [hunk 15]
```diff
@@ -98,6 +98,135 @@ waitForIoTAgent () {
...
+	                    "type": "Number"
+	                },
+	                {
+	                    "object_id": "digital_in_1",
+	                    "name": "flameDetected",
+	                    "type": "Number"
+	                },
+	                {
+	                    "object_id": "digital_in_2",
+	                    "name": "fillingLevelPercentage",
```
Comment (chicco785) [inline platform-stack/services]: fillingLevel

File: README_TTN.md [hunk 16]
```diff
@@ -103,6 +107,139 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+connect and used as password.
+Remember to remove the security code.
+
+```bash
+curl --location --request POST 'localhost:4042/iot/services' \
+--header 'fiware-service: openiot' \
+--header 'fiware-servicePath: /' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+  "services": [
```
Comment (chicco785) [inline README_TTN.md:124]: pretty print the json (some indenting is not aligned uniformly)

File: README_TTN.md [hunk 17]
```diff
@@ -103,6 +107,139 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+          "type": "Number"
+          },
+          {
+            "object_id": "digital_in_1",
+            "name": "flameDetected",
+            "type": "Number"
+            },
+            {
+              "object_id": "digital_in_2",
+              "name": "fillingLevelPercentage",
```
Comment (chicco785) [inline README_TTN.md]: fillingLevel

File: README_TTN.md [hunk 18]
```diff
@@ -103,6 +107,139 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+          "type": "Number"
+        },
+        {
+          "object_id": "relative_humidity_1",
+          "name": "relativeHumidity",
+          "type": "Number"
+          },
+          {
+            "object_id": "digital_in_1",
+            "name": "flameDetected",
```
Comment (chicco785) [inline README_TTN.md]: this is now part of status, no?

File: README_TTN.md (hunk 10 above)
Comment (chicco785) [inline README_TTN.md]: this is now part of status, no?

File: README_TTN.md (hunk 11 above)
Comment (chicco785) [inline README_TTN.md]: this is now part of status, no?

File: platform-stack/services [hunk 19]
```diff
@@ -98,6 +98,135 @@ waitForIoTAgent () {
...
+	                    "type": "Number"
+	                },
+	                {
+	                    "object_id": "relative_humidity_1",
+	                    "name": "relativeHumidity",
+	                    "type": "Number"
+	                },
+	                {
+	                    "object_id": "digital_in_1",
+	                    "name": "flameDetected",
```
Comment (chicco785) [inline platform-stack/services]: this is now part of status, no?

File: README_TTN.md [hunk 20]
```diff
@@ -103,6 +107,139 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+          "type": "Number"
+        },
+        {
+          "object_id": "digital_in_4",
+          "name": "CO",
+          "type": "Number"
+        },
+        {
+          "object_id": "digital_in_5",
+          "name": "airQuality",
```
Comment (chicco785) [inline README_TTN.md:208]: airQualityLevel

let's use numbers from international standards to compute the level:

https://www.airnow.gov/aqi/aqi-basics/

i.e. for no pollution, a number below 50 etc (to be fixed in arduino code)

File: platform-stack/services [hunk 21]
```diff
@@ -98,6 +98,120 @@ waitForIoTAgent () {
...
+	                    "type": "Number"
+	                }
+	                {
+	                    "object_id": "digital_in_1",
+	                    "name": "fillingLevel",
+	                    "type": "Number"
+	                }
+	                {
+	                    "object_id": "digital_in_2",
+	                    "name": "status",
```
Comment (chicco785) [inline platform-stack/services:135]: expression?

File: README_TTN.md [hunk 22]
```diff
@@ -103,6 +105,132 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+          "type": "Number"
+        },
+        {
+          "object_id": "digital_in_1",
+          "name": "fillingLevel",
+          "type": "Number"
+        },
+        {
+          "object_id": "digital_in_2",
+          "name": "status",
```
Comment (chicco785) [inline README_TTN.md:150]: expression?

File: README_TTN.md [hunk 23]
```diff
@@ -103,6 +105,132 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+          "type": "Number"
+        },
+        {
+          "object_id": "relative_humidity_1",
+          "name": "relativeHumidity",
+          "type": "Number"
+        },
+        {
+          "object_id": "digital_in_1",
+          "name": "flameDetected",
```
Comment (chicco785) [inline README_TTN.md]: remove?

File: README_TTN.md [hunk 24]
```diff
@@ -103,6 +105,132 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+        },
+        {
+          "object_id": "digital_in_1",
+          "name": "fillingLevel",
+          "type": "Number"
+        },
+        {
+          "object_id": "digital_in_2",
+          "name": "status",
+          "type": "Number"
```
Comment (chicco785) [inline README_TTN.md]: Text?

File: platform-stack/services [hunk 25]
```diff
@@ -98,6 +98,120 @@ waitForIoTAgent () {
...
+	                    "type": "Number"
```
Comment (chicco785) [inline platform-stack/services]: Text?

File: platform-stack/services (hunk 21 above)
Comment (casarf) [reply platform-stack/services:135]: Not working properly yet, there is something I'm missing

File: README_TTN.md (hunk 22 above)
Comment (chicco785) [reply README_TTN.md:150]: you marked as resolved, how was this resolved?

File: README_TTN.md (hunk 23 above)
Comment (chicco785) [reply README_TTN.md]: you marked as resolved, how was this resolved?

File: README_TTN.md (hunk 24 above)
Comment (chicco785) [reply README_TTN.md]: you marked as resolved, how was this resolved?

File: diorama/src/environment/cayenne_environment_seed_martel_002/cayenne_environment_seed_martel_002.ino [hunk 26]
```diff
@@ -174,12 +169,13 @@ void airQualitySetup() {
...
+    int value = sensor.getValue();
 
     if(DEBUG){
       Serial.print("Sensor value: ");
-      Serial.println(sensor.getValue());
-    }  
-    
+      Serial.println(value);
+    } 
+
```
Comment (chicco785) [inline diorama/src/environment/cayenne_environment_seed_martel_002/cayenne_environment_seed_martel_002.ino]: remove these spaces

File: README_TTN.md (hunk 20 above)
Comment (chicco785) [reply README_TTN.md:208]: i don't think this is solved looking at the code .ino

Comment (chicco785) [reply README_TTN.md:208]: ok, it's solved in the expression

File: platform-stack/services (hunk 12 above)
Comment (chicco785) [reply platform-stack/services]: password...

Comment (chicco785) [reply platform-stack/services]: for the passwords use an env variable that is rendered in the registration

File: platform-stack/services [hunk 27]
```diff
@@ -98,6 +98,128 @@ waitForIoTAgent () {
...
+	                {
+	                    "object_id": "digital_in_4",
+	                    "name": "CO",
+	                    "type": "Number"
+	                },
+	                {
+	                    "object_id": "digital_in_5",
+	                    "name": "airQuality",
+	                    "type": "Text",
+						"expression" : "digital_in_5 == 0 ? \"Good\" : digital_in_5 == 1 ? \"Low\" : digital_in_5 == 2 ? \"Med\" : \"High\" "
```
Comment (chicco785) [inline platform-stack/services]: align correctly

File: platform-stack/services [hunk 28]
```diff
@@ -98,6 +98,128 @@ waitForIoTAgent () {
...
+	                    "name": "UVvalue",
+	                    "type": "Number"
+	                }
+	            ],
+	            "internal_attributes": {
+	                "lorawan": {
+	                    "application_server": {
+	                        "host": "eu1.cloud.thethings.network",
+	                        "username": "oc-diorama-002@ttn",
+	                        "password": "NNSXS.WYCQ2VM2FNMXR5DNYYCYAHAQAACG4XA3FTBMRYA.SNSYCYYNLGVM4POZWSQ4VINZOUUPT5OAV3DU436W6AZXPRTLQCZA",
```
Comment (chicco785) [inline platform-stack/services]: it's a password... see comment above (about using env variable)

File: README_TTN.md (hunk 23 above)
Comment (chicco785) [reply README_TTN.md]: docs is not aligned with function in services

**Review by chicco785 — CHANGES_REQUESTED** [review]
see comments

File: README_TTN.md (hunk 24 above)
Comment (chicco785) [reply README_TTN.md]: docs is not aligned with function in services

File: README_TTN.md (hunk 22 above)
Comment (chicco785) [reply README_TTN.md:150]: docs is not aligned with function in services

File: platform-stack/services [hunk 29]
```diff
@@ -98,6 +98,128 @@ waitForIoTAgent () {
...
+	curl --location --request POST 'localhost:4042/iot/services' \
+	--header 'fiware-service: openiot' \
+	--header 'fiware-servicePath: /' \
+	--header 'Content-Type: application/json' \
+	--data-raw '{
+	    "services": [
+	        {
+				"device_id": "urn:ngsi-ld:WasteContainer:martel-ttn-001",
+     			"entity_name": "urn:ngsi-ld:WasteContainer:martel-ttn-001",
+				"device_id": "urn:ngsi-ld:WasteContainer:martel-ttn-001",
```
Comment (chicco785) [inline platform-stack/services]: duplicated

File: platform-stack/services [hunk 30]
```diff
@@ -98,6 +98,128 @@ waitForIoTAgent () {
...
+     			"entity_name": "urn:ngsi-ld:WasteContainer:martel-ttn-001",
```
Comment (chicco785) [inline platform-stack/services]: duplicated

File: README_TTN.md [hunk 31]
```diff
@@ -103,6 +105,141 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+curl --location --request POST 'localhost:4042/iot/services' \
+--header 'fiware-service: openiot' \
+--header 'fiware-servicePath: /' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+  "services": [
+    {
+      "device_id": "urn:ngsi-ld:AirQualityObserved:martel-ttn-002",
+      "entity_name": "urn:ngsi-ld:AirQualityObserved:martel-ttn-002",
+      "device_id": "urn:ngsi-ld:environment:martel-ttn-002",
```
Comment (chicco785) [inline README_TTN.md]: delete

File: README_TTN.md [hunk 32]
```diff
@@ -103,6 +105,141 @@ Details are available [here](https://blog.squix.org/2017/07/seeeduino-lora-gps-g
...
+      "entity_name": "urn:ngsi-ld:environment:martel-ttn-002",
```
Comment (chicco785) [inline README_TTN.md]: delete

File: platform-stack/services [hunk 33]
```diff
@@ -98,6 +98,135 @@ waitForIoTAgent () {
...
+	echo -e "\nSubscribing to mqtt\n"
+
+	curl --location --request POST 'localhost:4042/iot/services' \
+	--header 'fiware-service: openiot' \
+	--header 'fiware-servicePath: /' \
+	--header 'Content-Type: application/json' \
+	--data-raw '{
+	    "services": [
+	        {
+				"device_id": "urn:ngsi-ld:WasteContainer:martel-ttn-001",
```
Comment (chicco785) [inline platform-stack/services]: align

File: platform-stack/services [hunk 34]
```diff
@@ -98,6 +98,135 @@ waitForIoTAgent () {
...
+     			"entity_name": "urn:ngsi-ld:WasteContainer:martel-ttn-001",
+	            "entity_type": "WasteContainer",
+	            "apikey": "",
+	            "resource": "8CF957200005727C",
+				"expressionLanguage": "jexl",
```
Comment (chicco785) [inline platform-stack/services]: align

File: platform-stack/README.MD [hunk 35]
```diff
@@ -184,6 +184,7 @@ For more information on each service read the [Relevant Readings](#Relevant_Read
       ]
   }'
   ```
+>NOTE: In order to make the subscription work is needed to configure the .env file with correct passwords.
```
Comment (chicco785) [inline platform-stack/README.MD]: >NOTE: In order to make the subscription work is needed to configure the .env file with correct passwords.

=> 

> **NOTE:** For a correct deployment, copy `.env.example` to `.env` and configure it with the correct passwords for TTN.

**Review by chicco785 — CHANGES_REQUESTED** [review]

File: platform-stack/README.MD (hunk 35 above)
Comment (casarf) [reply platform-stack/README.MD]: Corrected

**Review by chicco785 — APPROVED** [review]

**Pull request merged**